proxy/limits/limits.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !windows
+
+// Package limits provides routines to check and enforce certain resource
+// limits on the Cloud SQL client proxy process.
+package limits
+
+import (
+	"fmt"
+	"syscall"
+
+	"github.com/GoogleCloudPlatform/cloudsql-proxy/logging"
+)
+
+var (
+	// For overriding in unittests.
+	getRlimitFunc = syscall.Getrlimit
+	setRlimitFunc = syscall.Setrlimit
+)
+
+// SetupFDLimits ensures that the process running the Cloud SQL proxy can have
+// at least wantFDs number of open file descriptors. It returns an error if it
+// cannot ensure the same.
+func SetupFDLimits(wantFDs uint64) error {
+	curr := &syscall.Rlimit{}
+	if err := getRlimitFunc(syscall.RLIMIT_NOFILE, curr); err != nil {
+		return fmt.Errorf("failed to read rlimit for max file descriptors: %v", err)
+	}
+
+	if curr.Max < wantFDs {
+		return fmt.Errorf("max FDs rlimit set to %d, cannot support 4K connections. If you are the system administrator, consider increasing this limit.", curr.Max)
```

Comment:
Should we try to set the limit to the Max so that we can at least try our hardest to increase this value? (I still think it's a good idea to print out this message when we can't update it to `wantFDs`.)

Reply:
I'm not quite sure I follow what you are asking here. We cannot increase the value past the Max returned by the current rlimits.

Reply:
Instead of returning an error when we can't set the rlimit to `wantFDs` (because it is greater than `curr.Max`), I'm thinking we log a message and set `wantFDs` to `curr.Max`.

Reply:
That was how I initially had it. Then I thought since the function is anyways returning an error, I'll just let the package return an error and let the caller deal with it. And the main function which calls this only logs the error, it does not fail the proxy for errors returned from here.

Let me know if you feel strongly though, I will change it.

Reply:
It looks like you have the proxy try to set the hard limit now, but that's not what I meant. I guess we could try setting the hard limit, but I think we should handle the case where users are not running the client proxy as root. Specifically, I'd like:

`--fd_rlimit=100` + `curr = {Cur: 10, Max: 50}` => `syscall.Setrlimit({Cur: 50, Max: 50})`

Feel free to ping me over chat if I'm being confusing still.